### PR TITLE
Support UpdateEvent for Update Operations

### PR DESCRIPTION
**Pull request by rdsr**
#926 Want feedback on the approach

TODO
1. Haven't added all events, want to discuss approach.
2. Some tests are failing, since `snapshotOperation` is null

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableOperations.java [hunk 1]
```diff
@@ -30,6 +30,8 @@
  */
 public interface TableOperations {
 
+  String tablePath();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableOperations.java:33]: Why not use `table.location()`? You can access that using `ops.current().location()`.

I'm really reluctant to add more methods to this interface since it is an important place where people can plug in their own customizations.

File: core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  private final String operation;
+  private final String tablePath;
+  private final Map<String, String> summary;
+  private final Iterable<String> addedFiles;
+  private final Iterable<String> appendedManifests;
+  private final Iterable<String> deletePaths;
+
+  public CreateSnapshotEvent(
+      String operation, String tablePath, Map<String, String> summary,
+      Iterable<String> addedFiles, Iterable<String> appendedManifests, Iterable<String> deletePaths) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java:34]: Why use table path/location instead of name? I think we should send a `name` field with whatever was used to load the table. We might need to add `name` to the `Table` interface to do that, but adding that method is easier than adding something to `TableOperations` because all the implementations are internal.

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java [hunk 3]
```diff
@@ -293,6 +294,8 @@ public void commit() {
     } catch (RuntimeException e) {
       LOG.warn("Failed to load committed table metadata, skipping manifest clean-up", e);
     }
+
+    Listeners.notifyAll(updateEvent());
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/SnapshotProducer.java:298]: This needs to be in a `try`/`catch` block to ensure that no exceptions are thrown. It is important for correctness that this returns successfully if the commit was successful. If the commit succeeds, but this throws an exception, then higher-level retries will re-attempt the operation on the same data and produce duplicates.

Also, if this is sent after commit, then it should include the sequence number that was committed.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 4]
```diff
@@ -317,6 +319,18 @@ private ManifestFile copyManifest(ManifestFile manifest) {
     }
   }
 
+  @Override
+  public UpdateEvent updateEvent() {
+    return new CreateSnapshotEvent(
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:324]: I think this should also include the snapshot ID and sequence number that were committed.

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java [hunk 5]
```diff
@@ -476,4 +476,6 @@ public Type primitive(Type.PrimitiveType primitive) {
     newFields.addAll(adds);
     return Types.StructType.of(newFields);
   }
+
+  //TODO: updateEvent
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/SchemaUpdate.java:480]: I think we can add update events individually.

File: api/src/main/java/org/apache/iceberg/events/UpdateEvent.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.events;
+
+public interface UpdateEvent {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/events/UpdateEvent.java:22]: What is the value of having an interface for all types of updates with no methods?

The listener code routes these events using `listeners.get(event.getClass())` so the interface doesn't help to subscribe to all events.

File: core/src/main/java/org/apache/iceberg/TableOperations.java (hunk 1 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/TableOperations.java:33]: I guess table location is also fine. We can parse out table name and db name from it.

File: core/src/main/java/org/apache/iceberg/SchemaUpdate.java (hunk 5 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/SchemaUpdate.java:480]: Makes sense

File: core/src/main/java/org/apache/iceberg/SnapshotProducer.java (hunk 3 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/SnapshotProducer.java:298]: Makes sense

File: core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java (hunk 2 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java:34]: I think we might need a reference to `Table` . Where we call commit and publish events we only have access to `TableOperations` . Any thoughts on how to get the `name` during `commits`?

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java:34]: Table creates all of these operations, so it can pass in its name as well as `ops`. I'd prefer that to trying to parse a location.

File: api/src/main/java/org/apache/iceberg/events/UpdateEvent.java (hunk 6 above)
Comment (rdsr) [reply api/src/main/java/org/apache/iceberg/events/UpdateEvent.java:22]: I was thinking that there may be some update operations where commit is being performed in their superclass. In those cases, we have to get the sub class' update event. Seeing through the code again, I don't see this and where a commit it being performed in a class it has all the information to send the event.  

I'll revert this change

Comment (rdsr) [reply api/src/main/java/org/apache/iceberg/events/UpdateEvent.java:22]: The issue is that I need to call the listeners after commit, but I don't have all the information to construct the event. For instance for `SnapshotProducer` where we are committing  I don't have information on how to construct the event and I'm relying on the subclass `MergingSnapshotProducer` to construct the right event properly

File: core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java (hunk 2 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java:34]: Maybe I can pass in the `name` as a property which can reside in table metadata. I can access that metadata during commit and I don't have to add any new api in `TableOperations` ?

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java:34]: I don't think we need to add to `TableOperations`. Table keeps track of its name, and is responsible for creating the operations classes, like `FastAppend`. So we can just pass the name in along side `TableOperations`:

```java
  @Override
  public AppendFiles newFastAppend() {
    return new FastAppend(name, ops);
  }
```

File: api/src/main/java/org/apache/iceberg/events/UpdateEvent.java (hunk 6 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/events/UpdateEvent.java:22]: Makes sense. We could also have `updateEvent()` return `Object`, since there are no requirements for objects passed to `Listeners.notifyAll`.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 7]
```diff
@@ -93,7 +93,7 @@ public Transaction newCreateTableTransaction(
     String baseLocation = location != null ? location : defaultWarehouseLocation(identifier);
     Map<String, String> tableProperties = properties != null ? properties : Maps.newHashMap();
     TableMetadata metadata = TableMetadata.newTableMetadata(schema, spec, baseLocation, tableProperties);
-    return Transactions.createTableTransaction(ops, metadata);
+    return Transactions.createTableTransaction(identifier.name(), ops, metadata);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:96]: Did you mean to use `identifier.toString` here as well?

File: core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java [hunk 8]
```diff
@@ -25,8 +25,8 @@
 
 public class BaseReplacePartitions
     extends MergingSnapshotProducer<ReplacePartitions> implements ReplacePartitions {
-  BaseReplacePartitions(TableOperations ops) {
-    super(ops);
+  BaseReplacePartitions(String tableName, TableOperations ops) {
+    super(tableName, ops);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:29]: Nit: Other places use `name`.

File: core/src/main/java/org/apache/iceberg/BaseTable.java [hunk 9]
```diff
@@ -39,6 +39,11 @@ public BaseTable(TableOperations ops, String name) {
     this.name = name;
   }
 
+  @Override
+  public String name() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTable.java:43]: Couldn't this method be added in a separate commit?

**Comment by rdsr** [comment]
@rdblue . Sorry I think this went in by mistake. This is not ready for review yet [Thought the final version would be somewhat similar]

File: core/src/main/java/org/apache/iceberg/PropertiesUpdate.java [hunk 10]
```diff
@@ -37,6 +37,7 @@
 import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT;
 
 class PropertiesUpdate implements UpdateProperties {
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/PropertiesUpdate.java:40]: Nit: this file doesn't need to change.

File: core/src/main/java/org/apache/iceberg/FastAppend.java [hunk 11]
```diff
@@ -49,7 +49,8 @@
   private ManifestFile newManifest = null;
   private boolean hasNewFiles = false;
 
-  FastAppend(TableOperations ops) {
+  FastAppend(String tableName, TableOperations ops) {
+    //TODO we use tableName to publish notification
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/FastAppend.java:53]: I think the append event should be created in this PR if we are sending events from `MergingSnapshotProducer`. I don't think it is necessary to implement events for `SnapshotManager` yet, though.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 12]
```diff
@@ -317,6 +321,22 @@ private ManifestFile copyManifest(ManifestFile manifest) {
     }
   }
 
+  @Override
+  public Object updateEvent() {
+    long snapshotId = snapshotId();
+    //long sequenceNumber = ops.refresh().snapshot(snapshotId).sequenceNumber();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:327]: Why comment this out? Looks correct to me.

File: core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+  public List<String> addedFiles() {
+    return addedFiles;
+  }
+
+  public List<String> appendedManifests() {
+    return appendedManifests;
+  }
+
+  public List<String> deletePaths() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java:77]: Minor: the other methods use past tense and refer to data files as "files", so this would be `deletedFiles()` right?

File: core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+  public Map<String, String> summary() {
+    return summary;
+  }
+
+  public List<String> addedFiles() {
+    return addedFiles;
+  }
+
+  public List<String> appendedManifests() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java:73]: Is this the set of manifests that were added to the dataset, or the ones that were passed to `appendManifest`? I think we should be careful that we know exactly what is in these messages. If it isn't obvious what these should be, then we should remove the data. We can always add more information later.

File: core/src/main/java/org/apache/iceberg/FastAppend.java (hunk 11 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/FastAppend.java:53]: So `FastAppend` extends `SnapshotProducer` . If I change `SnapshotProducer`, I'll have to change `BaseRewriteManifests`. OK to make that change in this PR?

**Comment by rdsr** [comment]
This is blocked on #968

File: core/src/main/java/org/apache/iceberg/BaseTable.java (hunk 9 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/BaseTable.java]: Removed

File: core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java (hunk 14 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java]: Removed for now.

File: core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java (hunk 13 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java]: fixed.

**Comment by rdsr** [comment]
This rb is ready for review again. As for #968 .  For now we can override `updateEvent()` to return a null. Later we can fix it as described here: https://github.com/apache/incubator-iceberg/pull/968#discussion_r415416694

File: core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java (hunk 8 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/BaseReplacePartitions.java:29]: I felt tableName was more appropriate. I have changed it as such in the code I added/modified. 
There are still uses of name though, e.g in Transaction api. I haven't changed them.  Let me know if u feel strong in using `name` or changing to `tableName` everywhere. I'm fine with either

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java (hunk 7 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java]: Fixed.

File: core/src/main/java/org/apache/iceberg/FastAppend.java (hunk 11 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/FastAppend.java:54]: Are you saying that `SnapshotProducer` would send the event? What needs to be changed?

File: core/src/main/java/org/apache/iceberg/BaseTable.java [hunk 15]
```diff
@@ -136,37 +136,37 @@ public RewriteManifests rewriteManifests() {
...
 
   @Override
   public DeleteFiles newDelete() {
-    return new StreamingDelete(ops);
+    return new StreamingDelete(name, ops);
   }
 
   @Override
-  public ExpireSnapshots expireSnapshots() {
+  public RemoveSnapshots expireSnapshots() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTable.java:153]: Why did this change? Can we not return the interface?

File: core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+  public long sequenceNumber() {
+    return sequenceNumber;
+  }
+
+  public Map<String, String> summary() {
+    return summary;
+  }
+
+  public List<String> addedFiles() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/events/CreateSnapshotEvent.java:67]: What happens if manifests are added? Is this incomplete?

**Comment by rdblue** [comment]
Thanks @rdsr, this looks mostly good, although I think the file lists in the event might be incomplete. I'd suggest removing them and just sending the summary, unless you have a use case for them.

I'd also like to get fast appends and snapshot manager changes done as soon as we can -- ideally in this PR. That way, we can guarantee that all changes to table contents produce events. Events aren't as useful unless you know you're getting all of the changes.

File: core/src/main/java/org/apache/iceberg/BaseTable.java (hunk 15 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/BaseTable.java:153]: Fixed. My bad.

**Comment by rdsr** [comment]
> I think the file lists in the event might be incomplete. I'd suggest removing them and just sending the summary, unless you have a use case for them.
I thin sending the summary is fine by me.

File: core/src/main/java/org/apache/iceberg/FastAppend.java (hunk 11 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/FastAppend.java:54]: Scratch that. I think this can be supported. I'll do that as part of this pr

**Comment by rdsr** [comment]
@rdblue . Thanks for the helpful review. I've address your comments and also added events for FastAppends and SnapshotsManager

File: core/src/main/java/org/apache/iceberg/FastAppend.java [hunk 17]
```diff
@@ -141,6 +144,18 @@ private ManifestFile copyManifest(ManifestFile manifest) {
     return newManifests;
   }
 
+  @Override
+  public Object updateEvent() {
+    long snapshotId = snapshotId();
+    long sequenceNumber = ops.refresh().snapshot(snapshotId).sequenceNumber();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/FastAppend.java:150]: The commit is going to update the current metadata, so there shouldn't be a need to `refresh` here. You'll get the correct information using `current`.

File: core/src/main/java/org/apache/iceberg/SnapshotManager.java [hunk 18]
```diff
@@ -127,6 +127,32 @@ public ManageSnapshots rollbackTo(long snapshotId) {
...
+  @Override
+  public Object updateEvent() {
+    if (targetSnapshotId == null) {
+      // NOOP operation, no snapshot created
+      return null;
+    }
+
+    switch (managerOperation) {
+      case ROLLBACK:
+        // rollback does not create a new snapshot
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/SnapshotManager.java:139]: It doesn't create a new snapshot, but should it send an event? Maybe as a follow-up?

**Comment by rdblue** [comment]
Overall, I think this is ready. I'd prefer to use `current` instead of `refresh` in the `updateEvent` methods, though. I'd probably wait to merge until then.

It would also be good to follow up with events in snapshot manager: rollback, fast-forward, and set current snapshot.

Last, I think that this is going to send events each time an individual operation is committed in a transaction. That's okay for now, but we may want to go back through and suppress events during a transaction and then send them all at once when the transaction succeeds. That will handle retries better.

File: core/src/main/java/org/apache/iceberg/SnapshotManager.java (hunk 18 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/SnapshotManager.java:139]: Yes. I'll create a followup for this. I presume this would be a specific Rollback event?

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/SnapshotManager.java:139]: Maybe a history modification event with a type that could be "rollback", "fast-forward", "cherry-pick", or "set-current"?

**Comment by rdsr** [comment]
Thanks @rdblue . I'll create followups for the remaining items

File: core/src/main/java/org/apache/iceberg/FastAppend.java (hunk 17 above)
Comment (rdsr) [reply core/src/main/java/org/apache/iceberg/FastAppend.java]: fixed
